shared/src/chain_state.rs

```diff
@@ -469,51 +476,68 @@ impl ChainState {
                     );
                 }
             } else {
-                tx_pool.enqueue_tx(entry.cycles, entry.size, entry.transaction);
+                let ret = self.pending_tx(tx_pool, entry.cycles, entry.size, entry.transaction);
+                if ret.is_err() {
+                    tx_pool.update_statics_for_remove_tx(entry.size, entry.cycles.unwrap_or(0));
+                    trace_target!(
+                        crate::LOG_TARGET_TX_POOL,
+                        "pending tx {:x} failed {:?}",
+                        tx_hash,
+                        ret
+                    );
+                }
             }
         }
     }
 
-    pub(crate) fn proposed_tx(
+    fn calculate_transaction_fee(&self, rtx: &ResolvedTransaction) -> Result<Capacity, PoolError> {
+        DaoCalculator::new(&self.consensus, self.store())
+            .transaction_fee(&rtx)
+            .map_err(|err| {
+                error_target!(
+                    crate::LOG_TARGET_TX_POOL,
+                    "Failed to generate tx fee for {:x}, reason: {:?}",
+                    rtx.transaction.hash(),
+                    err
+                );
+                PoolError::TxFee
+            })
+    }
+
+    #[allow(clippy::too_many_arguments)]
```

Comment:
It seems the function does not use `self`, why it is declared in `ChainState`? It's also a bad smell that the function name is `send_tx_to_pool`, it still needs a closure to tell the function how to add a transaction.

Reply:
I think the correct way to implement duplicated logic for different pool structures is defining some common operations as traits.

Reply:
It's hard to abstract common trait on a single pool, for example, https://github.com/nervosnetwork/ckb/blob/9aa1890062b97818e7eda8c82e2db39887966043/shared/src/tx_pool/pool.rs#L116

This is only for reusing the error handle, since we force verify txs before it enters a pool. Need further refractory to simplify this.